Feature: Reworked presence notification

ppenchev-storpool opened the pull request:
Rework the way the StorPool charms (cinder-storpool and storpool-block) notify their own peers and the other charm about the readiness of the units on specific machines and with specific StorPool IDs.

ppenchev-storpool committed:
Rework the presence hooks.

Make the storpool-presence and storpool-service interfaces send
the same data, a "storpool_presence" relation setting that contains
at least a "presence" hash with a version string and a data member.

Teach the storpool-block and cinder-storpool charms to exchange
presence information among their peers, only sharing it (through
their respective leaders) with the other charm if there are any
changes; also, share any changes received from the other charm with
all the peers.

ppenchev-storpool committed:
Add the get_cached_meta_config() method.

ppenchev-storpool committed:
Push the config down in the structure, always send it.

ppenchev-storpool committed:
Merge the new data into the current config.

This prevents e.g. a cinder-storpool peer sending a pretty much empty
presence data and overwriting the config just sent by the block leader.

ppenchev-storpool committed:
Move the presence data 'version' to the top level.

ppenchev-storpool committed:
QuasiConfig: do not require a default for get().

ryan-beisner commented:
Unit tests are failing on this branch:

```
pep8 create: /home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/.tox/pep8
pep8 installdeps: -r/home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/test-requirements.txt
pep8 installed: charmhelpers==0.18.5,charms.reactive==0.6.2,cliff==2.11.0,cmd2==0.8.1,extras==1.0.0,fixtures==3.0.0,flake8==3.5.0,future==0.16.0,Jinja2==2.10,linecache2==1.0.0,MarkupSafe==1.0,mccabe==0.6.1,mock==2.0.0,netaddr==0.7.19,os-testr==1.0.0,pbr==3.1.1,pkg-resources==0.0.0,prettytable==0.7.2,pyaml==17.12.1,pycodestyle==2.3.1,pyflakes==1.6.0,pyparsing==2.2.0,pyperclip==1.6.0,python-mimeparse==1.6.0,python-subunit==1.2.0,PyYAML==3.12,six==1.11.0,stestr==2.0.0,stevedore==1.28.0,Tempita==0.5.2,testtools==2.3.0,traceback2==1.4.0,unittest2==1.1.0,voluptuous==0.11.1
pep8 runtests: PYTHONHASHSEED='3361567810'
pep8 runtests: commands[0] | flake8 lib
pep8 runtests: commands[1] | flake8 --ignore=E402 unit_tests
unit_tests_35 create: /home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/.tox/unit_tests_35
unit_tests_35 installdeps: -r/home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/test-requirements.txt
unit_tests_35 installed: charmhelpers==0.18.5,charms.reactive==0.6.2,cliff==2.11.0,cmd2==0.8.1,extras==1.0.0,fixtures==3.0.0,flake8==3.5.0,future==0.16.0,Jinja2==2.10,linecache2==1.0.0,MarkupSafe==1.0,mccabe==0.6.1,mock==2.0.0,netaddr==0.7.19,os-testr==1.0.0,pbr==3.1.1,pkg-resources==0.0.0,prettytable==0.7.2,pyaml==17.12.1,pycodestyle==2.3.1,pyflakes==1.6.0,pyparsing==2.2.0,pyperclip==1.6.0,python-mimeparse==1.6.0,python-subunit==1.2.0,PyYAML==3.12,six==1.11.0,stestr==2.0.0,stevedore==1.28.0,Tempita==0.5.2,testtools==2.3.0,traceback2==1.4.0,unittest2==1.1.0,voluptuous==0.11.1
unit_tests_35 runtests: PYTHONHASHSEED='3361567810'
unit_tests_35 runtests: commands[0] | ostestr
{0} unit_tests.test_config.TestConfig.test_drop_cache [0.001263s] ... ok
{0} unit_tests.test_config.TestConfig.test_get_cached_dict [0.003025s] ... ok
{0} unit_tests.test_config.TestConfig.test_get_dict [0.000765s] ... ok
{0} unit_tests.test_service.TestStorPoolService.test_add_present_node [0.001735s] ... FAILED

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    b'Traceback (most recent call last):'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/.tox/unit_tests_35/lib/python3.5/site-packages/mock/mock.py", line 1305, in patched'
    b'    return func(*args, **keywargs)'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/unit_tests/test_service.py", line 276, in test_add_present_node'
    b'    list(map(lambda rid: [rid, jdata], rels)))'
    b'  File "/usr/lib/python3.5/unittest/case.py", line 820, in assertEqual'
    b'    assertion_func(first, second, msg=msg)'
    b'  File "/usr/lib/python3.5/unittest/case.py", line 1018, in assertListEqual'
    b'    self.assertSequenceEqual(list1, list2, msg, seq_type=list)'
    b'  File "/usr/lib/python3.5/unittest/case.py", line 1000, in assertSequenceEqual'
    b'    self.fail(msg)'
    b'  File "/usr/lib/python3.5/unittest/case.py", line 665, in fail'
    b'    raise self.failureException(msg)'
    b'AssertionError: Lists differ: [] != [[\'peer-rel/1\', \'{"here": {"new-node": "13[50 chars]}}\']]'
    b''
    b'Second list contains 2 additional elements.'
    b'First extra element 0:'
    b'[\'peer-rel/1\', \'{"here": {"new-node": "13"}}\']'
    b''
    b'- []'
    b'+ [[\'peer-rel/1\', \'{"here": {"new-node": "13"}}\'],'
    b'+  [\'peer-rel/42\', \'{"here": {"new-node": "13"}}\']]'
    b''
{0} unit_tests.test_service.TestStorPoolService.test_get_state [0.000740s] ... ok
{0} unit_tests.test_service.TestStorPoolService.test_init [0.000384s] ... ok
{0} unit_tests.test_service.TestStorPoolService.test_update_state [0.002770s] ... ok
{0} unit_tests.test_states.TestStates.test_get_set [0.002833s] ... ok
{0} unit_tests.test_states.TestStates.test_handle_events [0.000564s] ... ok
{0} unit_tests.test_states.TestStates.test_register_unregister [0.000486s] ... ok

==============================
Failed 1 tests - output below:
==============================

unit_tests.test_service.TestStorPoolService.test_add_present_node
-----------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    b'Traceback (most recent call last):'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/.tox/unit_tests_35/lib/python3.5/site-packages/mock/mock.py", line 1305, in patched'
    b'    return func(*args, **keywargs)'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/unit_tests/test_service.py", line 276, in test_add_present_node'
    b'    list(map(lambda rid: [rid, jdata], rels)))'
    b'  File "/usr/lib/python3.5/unittest/case.py", line 820, in assertEqual'
    b'    assertion_func(first, second, msg=msg)'
    b'  File "/usr/lib/python3.5/unittest/case.py", line 1018, in assertListEqual'
    b'    self.assertSequenceEqual(list1, list2, msg, seq_type=list)'
    b'  File "/usr/lib/python3.5/unittest/case.py", line 1000, in assertSequenceEqual'
    b'    self.fail(msg)'
    b'  File "/usr/lib/python3.5/unittest/case.py", line 665, in fail'
    b'    raise self.failureException(msg)'
    b'AssertionError: Lists differ: [] != [[\'peer-rel/1\', \'{"here": {"new-node": "13[50 chars]}}\']]'
    b''
    b'Second list contains 2 additional elements.'
    b'First extra element 0:'
    b'[\'peer-rel/1\', \'{"here": {"new-node": "13"}}\']'
    b''
    b'- []'
    b'+ [[\'peer-rel/1\', \'{"here": {"new-node": "13"}}\'],'
    b'+  [\'peer-rel/42\', \'{"here": {"new-node": "13"}}\']]'
    b''


======
Totals
======
Ran: 10 tests in 0.6350 sec.
 - Passed: 9
 - Skipped: 0
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 1
Sum of execute time for each test: 0.0146 sec.

==============
Worker Balance
==============
 - Worker 0 (10 tests) => 0:00:00.056445

Slowest Tests:
Test id                                                            Runtime (s)
-----------------------------------------------------------------  -----------
unit_tests.test_config.TestConfig.test_get_cached_dict             0.003
unit_tests.test_states.TestStates.test_get_set                     0.003
unit_tests.test_service.TestStorPoolService.test_update_state      0.003
unit_tests.test_service.TestStorPoolService.test_add_present_node  0.002
unit_tests.test_config.TestConfig.test_drop_cache                  0.001
unit_tests.test_config.TestConfig.test_get_dict                    0.001
unit_tests.test_service.TestStorPoolService.test_get_state         0.001
unit_tests.test_states.TestStates.test_handle_events               0.001
unit_tests.test_states.TestStates.test_register_unregister         0.000
unit_tests.test_service.TestStorPoolService.test_init              0.000
ERROR: InvocationError: '/home/ubuntu/git/storpool-charms/storpool-charms/layers/layer-storpool-helper/.tox/unit_tests_35/bin/ostestr'
___________________________________ summary ____________________________________
  pep8: commands succeeded
ERROR:   unit_tests_35: commands failed
```

thedac commented on lib/spcharms/service_hook.py:
```diff
@@ -58,80 +56,184 @@ def update_state(db, state, changed, key, name, value):
 
 
-def add_present_node(name, value, rel_name, rdebug=lambda s: s):
+def add_present_node(service, name, value, rdebug=lambda s: s):
```
Keep in mind unitdata is on a per-unit basis. Such that unit/0 will have different values from unit/1. In order to share that data with all the units either leader-set or relation-set (on the peer relation) must used.

thedac reviewed (commented):
Two points here.
One, it seems like it is mixing unitdata with relation data. Unitdata is very specific to the unit running the hook. It has no knowledge of its peers or its relationships. I just want to make sure there are no assumptions about what any given unit will have in unitdata at any given hook execution.

Two, the version number hardcoding is likely to make upgrades difficult.

ppenchev-storpool committed:
Fix the add_present_node() unit test.

The add_present_node() function no longer sends any data along
any relations; this is now done by send_presence_data(), which
deserves a unit test of its own.

ppenchev-storpool committed:
Merge the Travis CI config into the feature-presence branch.

ppenchev-storpool commented:
Hi,

The tests pass now.

The unit/relation data is a weird one, I admit.  Last week I replied to an e-mail with the subject "Storpool Charms Review - March 2018" where I explained the reasons for this implementation: basically, the storpool-block and cinder-storpool charms need to communicate across a container/host boundary, so the only way I thought of to do it was to keep all the data about a unit's peers, then have the leader pass it along to the other charm, then have the other charm's units distribute it among themselves.  As noted in that e-mail (there are some more details there, too), if there is a better way to do that, please let me know.

The version number is actually designed to make upgrades easier: it is a semver-like format version.  If there are changes to the format that do not remove existing fields, the version will remain 1.x and the older charms will still be able to parse the data before they are upgraded.  If there is a breaking change, the newer charms will send out data with a 2.x version and the older charms will (for the moment) simply ignore it until they are upgraded.

Best regards,
Peter

ppenchev-storpool committed:
repo: install a later version if available.

Until now, unless a specific version was requested, install_packages()
would not upgrade an already-installed package even if a newer version
was available in the repositories.  To support automated upgrades, this
needs to change.

ajkavanagh commented:
Note that the tests are hardwired to python3.5 ... however, moving forward it would be useful to wire them up for just python3 as python3.6 ships with Artful/Bionic, etc.  e.g.

`tox.ini`:

```ini
[tox]
envlist = pep8,unit_tests_3
skipsdist = True

[testenv:unit_tests_3]
basepython = python3
deps = -r{toxinidir}/test-requirements.txt
commands = ostestr {posargs}

[testenv:pep8]
basepython = python3
deps = -r{toxinidir}/test-requirements.txt
commands =
  flake8 {posargs} lib
  flake8 --ignore=E402 {posargs} unit_tests
```

Unit tests pass with these settings on artful. (17.10)
